core/src/main/scala/chisel3/package.scala

```diff
@@ -210,26 +211,140 @@ package object chisel3 {
+        var iter = 0
+        var curr_start = 0
+        val buf = mutable.ListBuffer.empty[Option[Printable]]
+        while(iter < s.size) {
+          if(s(iter) == '%' ) {
+            if(iter >= s.size - 1 || s(iter + 1 ) != '%') {
+              throw new UnknownFormatConversionException("Un-escaped % found")
+            }
+            if(curr_start < iter) {
+              buf += Some(PString(s.substring(curr_start,iter)))
+            }
+            buf += Some(Percent)
+            curr_start = iter + 2
+            iter += 2
+          }
+          else {
+            iter += 1
+          }
+        }
+
+        // Handle the tail
+        if(curr_start < iter) {
+          buf += Some(PString(s.substring(curr_start,iter)))
+        }
+        buf.toSeq
       }
+
+      sc.checkLengths(args) // Enforce sc.parts.size == pargs.size + 1
       val parts = sc.parts.map(StringContext.treatEscapes)
-      // Zip sc.parts and pargs together ito flat Seq
-      // eg. Seq(sc.parts(0), pargs(0), sc.parts(1), pargs(1), ...)
+
+      // The 1st part is assumed never to contain a format specifier.
+      // If the 1st part of a string is an argument - then the 1st part will be an empty String.
+      // So we need to parse parts following the 1st one to get the format specifiers if any
+      val partsAfterFirst = parts.slice(1, parts.size)
+
+      // Align parts to their potential specifiers
+      val partsAndSpecifierSeq = partsAfterFirst.zip(args).map {
+        case (part, arg) => {
+
+          // Check if part starts with a format specifier (with % - disambiguate with literal % checking the next character if needed to be %)
+          // In the case of %f specifier there is a chance that we need more information - so capture till the 1st letter (a-zA-Z).
+          // Example cf"This is $val%2.2f here" - parts - Seq("This is ","%2.2f here") - the format specifier here is %2.2f.
+          val idx_of_fmt_str =
+            if (
+              !part.isEmpty() && part.charAt(0) == '%' && (part.size == 1 || (part.size >= 2 && part.charAt(1) != '%'))
+            ) part.indexWhere { _.isLetter }
+            else -1
+
+          val fmt = if (idx_of_fmt_str >= 0) Some(part.substring(0, idx_of_fmt_str + 1)) else None
+          val fmtArgs: Printable = arg match {
+            case d: Data => {
+              fmt match {
+                case Some("%n") => Name(d)
+                case Some("%N") => FullName(d)
+                case Some("%s") => d.toString
+                case Some(fForm) if d.isInstanceOf[Bits] => FirrtlFormat(fForm.substring(1,2),d)
+                case Some(x) => {
+                  val msg = s"Illegal format specifier '$x' for Chisel Data type!\n"
+                  throw new UnknownFormatConversionException(msg)
+                }
+                case None => d.toPrintable
+              }
+            }
+            case p: Printable => {
+              fmt match {
+                case Some("%s") => p.toString()
```

Comment:
```suggestion
                case Some("%s") => PString(p.toString)
```
Do not include `()` on nullary methods that don't have side effects. This prints a warning (which is easy to miss in our sea of warnings).

Also better to not rely on the implicit conversion from `String` to `Printable`.